### PR TITLE
[FLINK-19083] Remove deprecated DataStream#split

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

It removes deprecated `DataStream#split`.

This was voted in https://lists.apache.org/thread.html/rc1cd927939fcdce3a2dca9f509b48be0740c50b73304c2d46c6a4347%40%3Cdev.flink.apache.org%3E

## Verifying this change

This change is mostly a trivial rework / code cleanup without any test coverage. A couple of tests were updated to use side outputs instead of the split/select.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java [hunk 1]
```diff
@@ -138,138 +132,6 @@ public void testBufferTimeout() {
 		}
 	}
 
-	/**
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java:141]: This one was not just testing "split"

**Review by aljoscha — COMMENTED** [review]
The change seems good to me. I had one comment about a removed test.

In general, did you find using side outputs ergonomic when you changed the tests?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java (hunk 1 above)
Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java]: Right, I missed the bit with testing partitioning pre union. I reverted and adjusted the test.

**Comment by dawidwys** [comment]
As for the question if it was easy to use side outputs instead of split/select. Personally I found it rather straightforward. In the IterateExample I even find the code much simpler, as the splitting is coupled within the mapper which looks more natural to me. There are two issues I found that could've been a bit easier:

1. It would be nice if we could have side outputs e.g. from RichMapFunction. I had to replace `MapFunction` with `ProcessFunction`
2. I also found a comment in the `OutputTag` a bit misleading: `It is also not possible to use generic type variables with OutputTags, such as 'Tuple2<A, B>'` It is actually possible to do, you just need to create the OutputTag as an anonymous class. I will change the comment in the exception.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 2]
```diff
@@ -567,28 +565,14 @@ private void addEdgeInternal(Integer upStreamVertexID,
...
-			StreamEdge edge = new StreamEdge(upstreamNode, downstreamNode, typeNumber, outputNames, partitioner, outputTag, shuffleMode);
+			StreamEdge edge = new StreamEdge(upstreamNode, downstreamNode, typeNumber,
+				partitioner, outputTag, shuffleMode);
 
 			getStreamNode(edge.getSourceId()).addOutEdge(edge);
 			getStreamNode(edge.getTargetId()).addInEdge(edge);
 		}
 	}
 
-	public <T> void addOutputSelector(Integer vertexID, OutputSelector<T> outputSelector) {
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java:577]: It seems we should also remove `addVirtualSelectNode` method, the related `virtualSelectNodes` map and its occurrence ?

**Comment by gaoyunhaii** [comment]
And for migrating from `split` to `side output`, I'm also think the source / async wait operator might need also support the side outputs since these special operators are not easy to be replaced with `ProcessFunctions`, and users will have to add a new operator to use the side output.

**Comment by aljoscha** [comment]
This looks good to merge! I think we can address @gaoyunhaii's comment as a follow-up, right?

**Review by aljoscha — APPROVED** [review]

**Pull request closed without merging**